python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py

```diff
@@ -1,70 +1,129 @@
 import operator
 from collections import defaultdict
 from datetime import datetime
-from typing import AbstractSet, Any, Callable, Dict, Iterable, Mapping, Optional, Set, Union, cast
+from functools import cached_property
+from typing import (
+    AbstractSet,
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Set,
+    Union,
+    cast,
+)
 
 from dagster import _check as check
 from dagster._core.definitions.partition import (
     PartitionsDefinition,
     PartitionsSubset,
 )
+from dagster._core.definitions.time_window_partitions import PartitionKeysTimeWindowPartitionsSubset
 from dagster._core.errors import (
     DagsterDefinitionChangedDeserializationError,
 )
 from dagster._core.instance import DynamicPartitionsStore
+from dagster._serdes.serdes import NamedTupleSerializer, whitelist_for_serdes
 
 from .asset_graph import AssetGraph
 from .events import AssetKey, AssetKeyPartitionKey
 
 
-class AssetGraphSubset:
-    def __init__(
-        self,
-        asset_graph: AssetGraph,
-        partitions_subsets_by_asset_key: Optional[Mapping[AssetKey, PartitionsSubset]] = None,
+class AssetGraphSubsetSerializer(NamedTupleSerializer):
+    def before_pack(self, value: "AssetGraphSubset") -> "AssetGraphSubset":
+        converted_partitions_subsets_by_serialized_asset_key = {}
+        for k, v in value.partitions_subsets_by_serialized_asset_key.items():
+            if isinstance(v, PartitionKeysTimeWindowPartitionsSubset):
```

Comment:
Would it make sense to put the custom serializer on `PartitionKeysTimeWindowPartitionsSubset` itself?

Reply:
It would be nice to not have to convert these subset objects everywhere they're used, but I think on principle objects that don't cross a serialization boundary should not be decorated with `@whitelist_for_serdes`.

Ideally we have some parallel entity that means "convert to Y, then serialize/deserialize Y". Implementing something like that adds additional scope I'd prefer to avoid at this time